File: examples/node/shutdown.js
/**
 * Runtime Shutdown Example - Graceful cleanup of all boxes.
 *
 * Demonstrates the runtime.shutdown() method:
 * - Graceful shutdown of all running boxes
 * - Custom timeout configuration
 * - Behavior after shutdown (operations fail)
 */

import { JsBoxlite } from '@boxlite-ai/boxlite';

async function main() {
  console.log('=== Runtime Shutdown Example ===\n');

  // Get the default runtime
  const runtime = JsBoxlite.withDefaultConfig();

  // Create a few boxes
  const boxes = [];
  for (let i = 0; i < 3; i++) {
    const box = await runtime.create({ image: 'alpine:latest' });
    boxes.push(box);
    console.log(`Created box ${i + 1}: ${box.id}`);
  }

  // Execute a simple command in each box
  for (let i = 0; i < boxes.length; i++) {
    const execution = await boxes[i].exec('echo', [`Hello from box ${i + 1}`]);
    const stdout = await execution.stdout();
    let line;
    while ((line = await stdout.next()) !== null) {
      console.log(`  Box ${i + 1}: ${line.trim()}`);
    }
    await execution.wait();
  }

  // Get metrics before shutdown
  const metrics = await runtime.metrics();
  console.log('\nBefore shutdown:');
  console.log(`  Running boxes: ${metrics.numRunningBoxes}`);
  console.log(`  Total commands: ${metrics.totalCommandsExecuted}`);

  // Shutdown with custom timeout (5 seconds)
  console.log('\nShutting down all boxes (5 second timeout)...');
  await runtime.shutdown(5);
  console.log('Shutdown complete!');

  // After shutdown, new operations will fail
  console.log('\nTrying to create a new box after shutdown...');
  try {
    await runtime.create({ image: 'alpine:latest' });
    console.log('ERROR: Expected this to fail!');
  } catch (e) {
    console.log(`Expected error: ${e.message}`);
  }

  console.log('\nDone!');
}

main().catch(console.error);

File: examples/c/shutdown.c
/**
 * Runtime Shutdown Example - Graceful cleanup of all boxes.
 *
 * Demonstrates the boxlite_runtime_shutdown() function:
 * - Graceful shutdown of all running boxes
 * - Custom timeout configuration
 * - Behavior after shutdown (operations fail)
 */

#include <stdio.h>
#include <stdlib.h>
#include "boxlite.h"

int main() {
    CBoxliteRuntime* runtime = NULL;
    CBoxliteError error = {0};
    BoxliteErrorCode code;

    printf("=== Runtime Shutdown Example ===\n\n");

    // Create runtime with default settings
    code = boxlite_runtime_new(NULL, NULL, &runtime, &error);
    if (code != Ok) {
        fprintf(stderr, "Failed to create runtime (code %d): %s\n",
                error.code, error.message ? error.message : "unknown");
        boxlite_error_free(&error);
        return 1;
    }

    // Create a few boxes
    const char* opts = "{\"rootfs\":{\"Image\":\"alpine:3.19\"},\"env\":[],\"volumes\":[],\"network\":\"Isolated\",\"ports\":[]}";

    CBoxHandle* boxes[3];
    for (int i = 0; i < 3; i++) {
        boxes[i] = NULL;
        error = (CBoxliteError){0};
        code = boxlite_create_box(runtime, opts, &boxes[i], &error);
        if (code != Ok) {
            fprintf(stderr, "Failed to create box %d (code %d): %s\n",
                    i + 1, error.code, error.message ? error.message : "unknown");
            boxlite_error_free(&error);
            continue;
        }
        char* id = boxlite_box_id(boxes[i]);
        printf("Created box %d: %s\n", i + 1, id);
        boxlite_free_string(id);
    }

    // Get metrics before shutdown
    char* metrics_json = NULL;
    error = (CBoxliteError){0};
    code = boxlite_runtime_metrics(runtime, &metrics_json, &error);
    if (code == Ok) {
        printf("\nBefore shutdown:\n");
        printf("  Metrics: %s\n", metrics_json);
        boxlite_free_string(metrics_json);
    } else {
        fprintf(stderr, "Failed to get metrics (code %d): %s\n",
                error.code, error.message ? error.message : "unknown");
        boxlite_error_free(&error);
    }

    // Shutdown with custom timeout (5 seconds)
    printf("\nShutting down all boxes (5 second timeout)...\n");
    error = (CBoxliteError){0};
    code = boxlite_runtime_shutdown(runtime, 5, &error);
    if (code != Ok) {
        fprintf(stderr, "Shutdown failed (code %d): %s\n",
                error.code, error.message ? error.message : "unknown");
        boxlite_error_free(&error);
    } else {
        printf("Shutdown complete!\n");
    }

    // After shutdown, new operations will fail
    printf("\nTrying to create a new box after shutdown...\n");
    CBoxHandle* new_box = NULL;
    error = (CBoxliteError){0};
    code = boxlite_create_box(runtime, opts, &new_box, &error);
    if (code == Ok && new_box) {
        printf("ERROR: Expected this to fail!\n");
        CBoxliteError stop_error = {0};
        boxlite_stop_box(new_box, &stop_error);
        boxlite_error_free(&stop_error);
    } else {
        printf("Expected error (code %d): %s\n",
               error.code, error.message ? error.message : "unknown");
        boxlite_error_free(&error);
    }

    // Clean up
    boxlite_runtime_free(runtime);

    printf("\nDone!\n");
    return 0;
}
